compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encoding C types is enabled.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // For for cross-language CFI support, the encoding must be compatible at the FFI
+                // boundary. For instance:
+                //
+                //     struct type1 {};
+                //     void foo(struct type1* bar) {}
+                //
+                // is encoded as:
+                //
+                //     !1 = !{i64 0, !"_ZTSFvP5type1E"}
+                //
+                // So, encode any repr(C) user-defined type [or extern types (i.e., ty::Foreign)] as
+                // <length><name>, where <name> is <unscoped-name>.
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                // Generalize any repr(transparent) user-defined type that is either a pointer or
+                // reference, and either references itself or any other type that contains or
+                // references itself, to avoid a reference cycle.
+                if ty0.is_any_ptr() && ty0.contains(ty) {
+                    s.push_str(&encode_ty(
+                        tcx,
+                        ty0,
+                        dict,
+                        options | EncodeTyOptions::GENERALIZE_POINTERS,
+                    ));
+                } else {
+                    s.push_str(&encode_ty(tcx, ty0, dict, options));
+                }
+            } else {
+                // u<length><name>[I<element-type1..element-typeN>E] as vendor extended type
+                s.push_str(&encode_ty_name(tcx, def_id));
+                if adt_def.is_struct() || adt_def.is_union() {
+                    let tys: Vec<Ty<'tcx>> = substs.types().map(|ty| ty).collect();
```

Comment:
This should eventually encode all kinds of substs, not just types. It would be required to distinguish between all pairs of types in the following example:

```rust
#![feature(adt_const_params)]
pub fn make<T>() -> T { loop {} }

pub struct Array<T, const N: usize>([T; N]);
#[derive(PartialEq, Eq)]
pub struct Color { r: u8, g: u8, b: u8 }
pub struct Widget<const C: Color>;

pub fn main() {
    make::<Array<i8, 8>>();
    make::<Array<u8, 8>>();

    make::<Array<usize, 20>>();
    make::<Array<usize, 30>>();

    make::<Widget<{Color { r: 0, g: 0, b: 255 }}>>();
    make::<Widget<{Color { r: 255, g: 255, b: 0 }}>>();

    make::<&dyn for<'a, 'b> Fn(&'a u32, &'b u32) -> &'a u32>();
    make::<&dyn for<'a, 'b> Fn(&'a u32, &'b u32) -> &'b u32>();
}
```

Reply:
Done. (We currently don't distinguish between the following:

```Rust
    make::<Widget<{Color { r: 0, g: 0, b: 255 }}>>();
    make::<Widget<{Color { r: 255, g: 255, b: 0 }}>>();
```
Because we don't encode structure members.)